0019/README.md

```diff
@@ -0,0 +1,146 @@
+---
+name: Endpoint Metadata Management Protocol
+shortname: 19/EPMMP
+status: raw
+editor: Vitalii Kozlovskyi <vkozlovskyi@kaaiot.io>
+---
+
+
+<!-- toc -->
+
+
+# Introduction
+
+Endpoint Metadata Management Protocol (EPMMP) is designed for managing endpoint metadata between Kaa services.
+
+EPMMP complies with the [Inter-Service Messaging](/0003/README.md) guidelines.
+
+
+# Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+The following terms and definitions are used in this RFC.
+
+- **Endpoint metadata repository (repository)**: any service that exposes the EPMMP interface to other services for managing endpoint metadata.
+
+- **Endpoint metadata management client (client)**: any service that uses exposed EPMMP interface to manage endpoint metadata.
+
+
+# Design
+
+## Endpoint metadata get request
+
+In order to request endpoint metadata, clients MUST [emit](/0003/README.md#targeted-messaging) `EndpointMetadataGetRequest` message to the following NATS subject:
+```
+kaa.v1.service.<ep-metadata-service-instance-name>.epmmp.ep-metadata-get-request
+```
+where:
+- `<ep-metadata-service-instance-name>` is the endpoint metadata repository service instance name.
+
+The NATS message payload is an Avro object with the following schema ([0019-endpoint-metadata-get-request.avsc](./0019-endpoint-metadata-get-request.avsc)):
```

Comment:
Let us follow the `16/ECAP` description...


*Endpoint metadata get request* message is a [targeted message](/0003/README.md#targeted-messaging) that the client sends to the repository to retrieve endpoint metadata.

The client MUST send endpoint metadata get request messages using the following NATS subject:
```
kaa.v1.service.{ep-metadata-service-instance-name}.epmmp.ep-metadata-get-request
```
where:
- `{ep-metadata-service-instance-name}` is the endpoint metadata repository service instance name.

The client MUST include NATS `replyTo` field to handle the response.
It is RECOMMENDED to follow the subject format described in [3/ISM session affinity section](/0003/README.md#session-affinity):
```
kaa.v1.replica.{client-service-replica-id}.epmmp.ep-metadata-get-response
```

Endpoint metadata get request message payload MUST be an [Avro-encoded](https://avro.apache.org/) object with the following schema ([0019-endpoint-metadata-get-request.avsc](./0019-endpoint-metadata-get-request.avsc)):

Reply:
There are issues with `kaa.v1.replica.{client-service-replica-id}.epmmp.ep-metadata-get-response` subject.
We should not only receive the answer, but also match the request to the response.
I think the right way to do it is the NATS "Inbox" feature. (In that case, 3/ISM needs to be updated)

Reply:
@ubombi It's just recommendation.

Reply:
@ubombi To match the response to the request the client implementation can use the message correlation ID. Or of course, these calls can be implemented with the NATS inbox feature.